src/wmtk/invariants/FunctionInvariant.hpp

```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <memory>
+#include "Invariant.hpp"
+
+namespace wmtk {
+
+namespace function {
+class PerSimplexFunction;
+}
+
+namespace invariants {
+
+class FunctionInvariant : public Invariant
+{
+public:
+    FunctionInvariant(
+        const PrimitiveType type,
+        const std::shared_ptr<function::PerSimplexFunction>& func);
+
+    bool after(
+        const std::vector<Tuple>& top_dimension_tuples_before,
+        const std::vector<Tuple>& top_dimension_tuples_after) const override;
+
+private:
+    std::shared_ptr<function::PerSimplexFunction> m_func;
```

Comment:
do you intend on having this function shrared between multiple places? if you're not aware, std::function has a check if it has a value or not https://en.cppreference.com/w/cpp/utility/functional/function/operator_bool

Reply:
yes it will be shared